Article to explain how to do Reverse Engineering on EVM contracts (#4581)
ethereum/ethereum-org-website: src/content/developers/tutorials/reverse-engineering-a-contract/index.md

---
title: "Reverse Engineering a Contract"
description: How to understand a contract when you don't have the source code
author: Ori Pomerantz
lang: en
sidebar: true
tags: ["evm", "opcodes", "reverse engineering", "decompiler"]
skill: advanced
published: 2021-12-30
---

## Introduction {#introduction}

*There are no secrets on the blockchain*, everything that happens is consistent, verifiable, and publicly available. Ideally, [contracts should have their source code published and verified on Etherscan](https://etherscan.io/address/0xb8901acb165ed027e32754e0ffe830802919727f#code). However, [that is not always the case](https://etherscan.io/address/0x2510c039cc3b061d79e564b38836da87e31b342f#code).

There are reverse compilers, but they don't always produce [usable results](https://etherscan.io/bytecode-decompiler?a=0x2510c039cc3b061d79e564b38836da87e31b342f). In this article you learn how to manually reverse engineer and understand a contract from [the opcodes](https://github.com/wolflo/evm-opcodes), as well as how to interpret the results of a decompiler.

> To be able to understand this article you should already know the basics of the EVM, and be at least somewhat familiar with EVM assembler. [You read about about these topics here](https://medium.com/mycrypto/the-ethereum-virtual-machine-how-does-it-work-9abac2b7c9e).

Reviewer: ```suggestion
To be able to understand this article you should already know the basics of the EVM, and be at least somewhat familiar with EVM assembler. [You can read about about these topics here](https://medium.com/mycrypto/the-ethereum-virtual-machine-how-does-it-work-9abac2b7c9e).
```

I purposely did not put in the heading what this handler does. The point isn't to teach you how this specific contract works, but how to reverse engineer contracts. You will learn what it does the same way I did, by following the code.

We get here from several places:
* If there is call data of 1, 2, or 3 bytes (from offset 0x63)
* If the method signature is unknown (from offsets 0x42 and 0x5D)

| Offset | Opcode | Stack |
| -: | - | - | 
| 7C | JUMPDEST |
| 7D | PUSH1 0x00   | 0x00
| 7F | PUSH2 0x009d | 0x9D 0x00
| 82 | PUSH1 0x03   | 0x03 0x9D 0x00
| 84 | SLOAD        | Storage[3] 0x9D 0x00

This is another storage cell, one that I couldn't find in any transactions so it's harder to know what it means. The code below will make it clearer.

| Offset | Opcode | Stack |
| -: | - | - | 
| 85 | PUSH20 0xffffffffffffffffffffffffffffffffffffffff | 0xff....ff Storage[3] 0x9D 0x00
| 9A | AND | Storage[3]-as-address 0x9D 0x00

These opcodes truncate the value we read from Storage[3] to 160 bits, the length of an Ethereum address.

| Offset | Opcode | Stack |
| -: | - | - | 
| 9B | SWAP1 | 0x9D Storage[3]-as-address 0x00
| 9C | JUMP  | Storage[3]-as-address 0x00

> This just is superflous, since we're going to the next opcode. This code isn't nearly as gas-efficient as it could be.

Reviewer: ```suggestion
This jump is superfluous, since we're going to the next opcode. This code isn't nearly as gas-efficient as it could be.
```

| B5 | RETURNDATACOPY | RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address

Here we copy all the return data to the memory buffer starting at 0x80. 

| Offset | Opcode | Stack |
| -: | - | - | 
| B6 | DUP2 | (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address |
| B7 | DUP1 | (((call success/failure))) (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address |
| B8 | ISZERO |  (((did the call fail))) (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address |
| B9 | PUSH2 0x00c0 | 0xC0 (((did the call fail))) (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address
| BC | JUMPI | (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address
| BD | DUP2  | RETURNDATASIZE (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address
| BE | DUP5  | 0x80 RETURNDATASIZE (((call success/failure))) RETURNDATASIZE (((call success/failure))) 0x80 Storage[3]-as-address
| BF | RETURN | |

>   
>   
>   
>   

Reviewer: ```suggestion
```

  

| Offset | Opcode | Stack |
| -: | - | - |
| 144	| JUMPDEST
| 145	| POP
| 146	| PUSH2 0x00da | 0xDA
| 149	| PUSH2 0x0153 | 0x0153 0xDA
| 14C	| CALLDATASIZE | CALLDATASIZE 0x0153 0xDA
| 14D	| PUSH1 0x04   | 0x04 CALLDATASIZE 0x0153 0xDA |
| 14F	| PUSH2 0x018f | 0x018F 0x04 CALLDATASIZE 0x0153 0xDA |
| 152	| JUMP         | 0x04 CALLDATASIZE 0x0153 0xDA |
| 18F	| JUMPDEST     | 0x04 CALLDATASIZE 0x0153 0xDA |
| 190	| PUSH1 0x00   | 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 192	| PUSH1 0x20   | 0x20 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 194	| DUP3         | 0x04 0x20 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 195	| DUP5         | CALLDATASIZE 0x04 0x20 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 196	| SUB          | CALLDATASIZE-4 0x20 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 197	| SLT          | CALLDATASIZE-4<32 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 198	| ISZERO       | CALLDATASIZE-4>=32 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 199	| PUSH2 0x01a0 | 0x01A0 CALLDATASIZE-4>=32 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 19C	| JUMPI        | 0x00 0x04 CALLDATASIZE 0x0153 0xDA |

> It looks like this function takes at least 32 bytes (one word) of call data. If it doesn't get it:
> 
>   
> | Offset | Opcode | Stack |
> | -: | - | - |  
> | 19D | DUP1 | 0x00 0x00 0x04 CALLDATASIZE 0x0153 0xDA | 
> | 19E	| DUP2 | 0x00 0x00 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
> | 19F	| REVERT  
>   
> The transaction is reverted without any return data.

Reviewer: Felt this read a little funny... how about this:
```suggestion
It looks like this function takes at least 32 bytes (one word) of call data.

| Offset | Opcode | Stack |
| -: | - | - |  
| 19D | DUP1 | 0x00 0x00 0x04 CALLDATASIZE 0x0153 0xDA | 
| 19E	| DUP2 | 0x00 0x00 0x00 0x04 CALLDATASIZE 0x0153 0xDA |
| 19F	| REVERT  

If it doesn't get the call data the transaction is reverted without any return data.
```
